benchexec/resources.py

```diff
@@ -210,118 +218,738 @@ def _get_cpu_cores_per_run0(
+            book = get_book_id_for_core(core)
+            cores_of_book[book].append(core)
+    except FileNotFoundError:
+        cores_of_book = {}
+        logging.debug(
+            "No book information available at /sys/devices/system/cpu/cpuX/topology/"
+        )
+    logging.debug("Books of cores are %s.", cores_of_book)
+    return cores_of_book
+
+
+def get_drawer_id_for_core(core: int) -> int:
+    """Get the id of the drawer a core belongs to."""
+    return int(util.read_file(f"/sys/devices/system/cpu/cpu{core}/topology/drawer_id"))
+
+
+def get_drawer_mapping(allCpus_list: List[int]) -> HierarchyLevel:
+    cores_of_drawer = collections.defaultdict(list)
+    try:
+        for core in allCpus_list:
+            drawer = get_drawer_id_for_core(core)
+            cores_of_drawer[drawer].append(core)
+    except FileNotFoundError:
+        cores_of_drawer = {}
+        logging.debug(
+            "No drawer information available at /sys/devices/system/cpu/cpuX/topology/"
+        )
+    logging.debug("drawers of cores are %s.", cores_of_drawer)
+    return cores_of_drawer
+
+
+def get_L3cache_id_for_core(core: int) -> int:
+    """Check whether index level 3 is level 3 cache"""
+    dir_path = f"/sys/devices/system/cpu/cpu{core}/cache/"
+    index_L3_cache = ""
+    for entry in os.listdir(dir_path):
+        if entry.startswith("index"):
+            if (
+                int(
+                    util.read_file(
+                        f"/sys/devices/system/cpu/cpu{core}/cache/{entry}/level"
                     )
-                if len(cores) >= coreLimit:
-                    break
-            cores_with_siblings.update(cores)
-            cores = cores[:coreLimit]  # shrink if we got more cores than necessary
-            # remove used cores such that we do not try to use them again
-            cores_of_unit[unit] = [
-                core for core in cores_of_unit[unit] if core not in cores
-            ]
-
-        assert len(cores) == coreLimit, (
-            f"Wrong number of cores for run {run + 1} of {num_of_threads} "
-            f"- previous results: {result}, "
-            f"remaining cores per CPU/memory region: {cores_of_unit}, "
-            f"current cores: {cores}"
+                )
+                == 3
+            ):
+                index_L3_cache = entry
+                break
+    """Get the id of the Level 3 cache a core belongs to."""
+    return int(
+        util.read_file(f"/sys/devices/system/cpu/cpu{core}/cache/{index_L3_cache}/id")
+    )
+
+
+def get_L3cache_mapping(allCpus_list: List[int]) -> HierarchyLevel:
+    cores_of_L3cache = collections.defaultdict(list)
+    try:
+        for core in allCpus_list:
+            L3cache = get_L3cache_id_for_core(core)
+            cores_of_L3cache[L3cache].append(core)
+    except FileNotFoundError:
+        cores_of_L3cache = {}
+        logging.error(
+            "Level 3 cache information not available at /sys/devices/system/cpu/cpuX/cache/cacheX"
```

Comment:
Why is this an error? An error means that BenchExec cannot continue to work. Is this the case? If yes, then an exception should be thrown. If not, then it should not be an error message.

Is this something important for the user? If not, then a debug message should be used.

Reply:
Changed into a debug message